### DISCUSSION TITLE
Keyboard Shortcuts.

### DISCUSSION THREAD

**Discussion by EmilioRoberto** [soundshed/soundshed-app#7] [Ideas]
I hope there is a value in having endusers in this group to ask questions. I promise to try not to be annoying.

Part of what i do is edit video, and I find that keyboard shortcuts are useful. Would it be possible to add keyboard shortcuts to adjust levels? I imagine the use of the software where the player is not right next to the amp and wants to quickly adjust the tone or volume up or down and having certain keys on the keyboard raise and lower the gain, master volume, high, mid and low might be helpful.

**Comment by EmilioRoberto** [comment]
I see that when I am clicked on a knob, I can raise and lower the level with the up and down keys, but what I am referring to would be more like just the "T" and "G" raise and lower the main volume etc.

**Reply by webprofusion-chrisc** [reply]
Thanks, I'm sure it's possible, interesting idea.

 The audio controls UI knob is a component I didn't write, so the current keyboard controls come without any effort from me.

**Comment by crick616** [comment]
I'll add my vote for keyboard shortcuts. I'm a dinosaur IT guy who learned computers on CP/M (you might have to look that up... most people have never heard of it) long before mice, touchpads, Windows, etc. I strive to learn keyboard shortcuts in all software in order to keep my fingers where they belong - on the keyboard. And I'm usually faster than any mouser because of it.

**Reply by webprofusion-chrisc** [reply]
I've actually used a version of CP/M on an Amstrad CPC6128, it was probably 35 years ago though :)

So for keyboard shortcuts, what should they be and what should they do?

**Comment by crick616** [comment]
Wow,,, a fellow dinosaur! There aren't many of us left since the meteor.  OK, I'm just spitballing ideas here. I would like to see something like combo-key shortcuts to select a particular control and then after selecting it, use the up/down arrow keys to adjust that control (and then followed by ESC to de-select the control).

Controls could be selected by:
 Bass: ctl-B
 Mid: ctl-M
 Treble: ctl-T
 Mod: ctl-C  (would have chosen "M"od but it was already taken by "M"id, so "C"horus or "F"lange)
 Delay: ctl-D
 Reverb: ctl-R
Other shortcuts...
 Select Preset 1: ctl-1    Save current settings to Preset 1: alt-1 (or maybe ctl-alt-1 to minimize chance of accidental over-write)
 Select Preset 2: ctl-2    Save current settings to Preset 2: alt-2   "
 Select Preset 3: ctl-3    Save current settings to Preset 3: alt-3   "
 Select Preset 4: ctl-4    Save current settings to Preset 4: alt-4   "

**Reply by crick616** [reply]
Today I installed your app and spent some quality time with it. I really like the display and visibility of the settings. I'm now rethinking my desire for the shortcuts I mentioned yesterday. The ease of using your app as designed is very good and I don't know if I would even use the shortcuts. 

The things I would hope to see added are controls for Gain, Master, Output, Music Volume, and perhaps Tap. Also, will there eventually be a way to save settings to presets 1-4?

I absolutely prefer using your PC app instead of my phone or tablet. It's much easier to see and control with a laptop and I (fortunately) have one I can dedicate to this purpose. This is EXACTLY what I hoped for. Thank you for developing it.

**Reply by webprofusion-chrisc** [reply]
Thanks, glad you like it! This is what the (currently hidden because it's not wired up) device controls section looks like:
![image](https://user-images.githubusercontent.com/2445502/110282074-cf466680-8018-11eb-9e0c-06227e4c9f73.png)

Note that the amp level output control (the real volume control) and the music volume control are both not automated via bluetooth so you can only turn those manually, you can adjust bluetooth audio by turning up your PC volume though.

I will eventually add the option to store a preset to a hardware channel I just haven't gotten around to that yet. As a workaround you can just long-press the hardware channel button on the amp itself to store what you currently have.

**Comment by foebyeahright** [comment]
I had a similar Idea...but for a different reason:
I want to try to build a cheap wired footcontroller out of an old keyboard. Strip it of all keys but 4 and then add big "stompplates" to these 4 keys for channel selection. Since this is a complete different use of "hotkeys" i think it would be nice to user-customize these.. so i can spread the keys I need nicely across the keyboard...

**Reply by lavabyrd** [reply]
I'm happy to look into setting up some of these if you wish. Perhaps the basic 1,2,3,4 to select the presets and based on that, we could look at adding more?

**Reply by foebyeahright** [reply]
Sweet

**Reply by lavabyrd** [reply]
It'll depend on @webprofusion-chrisc - I'm not sure if he has other plans around keyboard shortcuts and I don't want to toe step :)

**Reply by webprofusion-chrisc** [reply]
@lavabyrd sure thing! We maybe need a new settings tab then in there key mapping to actions (key x maps to action y) would be one section. Eventually other settings we may eventually have might include accounts (soundshed.com & PG, possibly in the future soundcloud etc), midi mapping (event x maps to action y).

**Reply by lavabyrd** [reply]
Good shout. I'll get started on it over the next few days and we can have a look

**Reply by foebyeahright** [reply]
Midimapping +1 👍 

Never thought of that but thats probably the "best" solution...

But I'd be really happy with hotkeys :D

**Reply by webprofusion-chrisc** [reply]
Yeah the midi mapping idea is kind of an extension to the same idea, where we define a set of actions like selecting specific presets then could have several different ways of triggering those. 

The benefit of keyboard and midi mapping is you could map to not just the 4 hardware presets but to any saved presets - I've not really thought it through but we don't need to limit to the 4 hardware presets. I don't know if sending presets is slower than switching the hardware preset mind you.

**Reply by foebyeahright** [reply]
Hey! Was just wondering if u guys had the time to work on the keyboard ahortcut thingy :)

**Reply by webprofusion-chrisc** [reply]
Nothing from me recently, I've been working on allowing the video lessons to keep playing when you switch tab - other than that not much.

**Reply by lavabyrd** [reply]
Personally I haven't had time but I'm hoping to get to it this weekend!

**Reply by foebyeahright** [reply]
Both good news though :)

**Reply by foebyeahright** [reply]
...been a while... Just wondering if any progress has been made... Got a week off the week after next week and i intend to mess around with sparky a bit :D

**Reply by webprofusion-chrisc** [reply]
@foebyeahright not from me, I've been heads down with work stuff. The beauty of open source is that we all get to watch code grow old together :) I'm sure I'll get back into this but it could be weeks or months.

**Reply by foebyeahright** [reply]
Thanks for the info :)

**Comment by jhilleary** [comment]
This is great! I have an Airstep Lite pedal with the ability to map to any keyboard inputs via BT MIDI, so this would make it possible to finally use a footswitch for the presets and other effects without modding!

**Reply by webprofusion-chrisc** [reply]
From my understanding the Airstep Lite can also work with the official Spark app by pretending to be a keyboard input device or something like that? I presume you've tried that already.

**Reply by jhilleary** [reply]
I’ve seen that video on the FB mod page, but I tried everything I could and couldn’t get it to work. The closest I got was being able to Tab to select different menus and scroll/select, but couldn’t get to selecting options that matter.

**Reply by webprofusion-chrisc** [reply]
Oh I also posted this just now: https://sporkdemo.soundshed.com/ it's a web bluetooth thing (chrome/edge) which can talk to the spark without extra software. You can press keys 1-4 as well.

**Reply by jhilleary** [reply]
Awesome! I’ll try it out later on today when I get a chance

**Reply by jhilleary** [reply]
Success! Worked for me straight away, well done! Mapped my Airstep and boom, got my presets! Having shortcuts for all pedals and effects would be the bomb, at least for me and possibly others with a similar solution in mind. Part of my day job is QA testing .Net apps, so happy to pitch in with testing when I can.

**Reply by webprofusion-chrisc** [reply]
Great!

**Reply by foebyeahright** [reply]
this is really cool. Is there any way to have this (just the preset selection) as a standalone app that runs in background? So I could use my sheetmusic in the foreground (or a backing track or whatever) and switch sounds in the background?

**Reply by webprofusion-chrisc** [reply]
It probably depends what mechanism you want to use to switch, is it a midi pedal or something that acts like a (computer) keyboard? Midi controller is probably easier to do in the background.

**Reply by foebyeahright** [reply]
Computer Keyboard would be my nr. One choice....actually i am still hoping for the hotkeyimplementation to soundshed :D

**Reply by webprofusion-chrisc** [reply]
The problem with keyboard is that the foreground app gets the input, so if soundshed wasn't the foreground app then the keystrokes would go to the other app instead. There are ways around that (global hotkeys) but they're potentially complex and so they're a little unlikely to happen.

**Reply by foebyeahright** [reply]
Ahhh... I see the troubles....well... any chance to get hotkeys to soundshed anyway? Would love to see that  :)

**Reply by lavabyrd** [reply]
👋🏻 Life's been a little mad for me so I haven't gotten near this but i _think_ i may be able to do system wide hotkeys as well as in-app but I don't have an ETA for when that may be. It absolutely won't be this week anyway I'm afraid

**Comment by geppo1988** [comment]
Hi all :) news about this? Was it integrated somehow the channel selection using keystrokes? Thanks!!!

**Reply by webprofusion-chrisc** [reply]
Hi, no I haven't been working on this, it would be great if we could encourage more developers to get involved. I've not even been playing guitar recently :)

**Reply by webprofusion-chrisc** [reply]
Now working on this and midi mapping at the same time.

**Comment by webprofusion-chrisc** [comment]
The first draft of a feature for input mappings has now been added. This allows midi and (computer) keyboard inputs to be mapped to amp channel selections. Currently keys 1-4 are mapped to amp channels, Midi notes C3 to F3 are also mapped to channels. A learn function (to train different keys/inputs) is still planned.

**Reply by gfm2000** [reply]
I'm new to GitHub and this app, so please forgive any stuff-ups I make here. I just stumbled across this app and it is amazing. I am visually impaired and the Spark app on mobile is pretty useless to me (and PG weren't really interested in hearing about it's lack of accessibility).
This app, for the most part, is the perfect companion for the amp. Only things that aren't accessible with my screenreader are the adjustment controls, but I'm okay with that (though it would be nice to have them readable!).
As far as keyboard shortcuts go, what would be great is if the effects can have shortcuts assigned to them (as mentioned in a previous post). I bought a USB footswitch for my Boss Katana and am doing the same thing using an app on Android. Being able to toggle boosters and mods with the tap of a foot (or finger on a keyboard) would be ideal. If this is already there, then I must have missed it.
Keep up the great work. This app has made my whole Spark experience more usable (and enjoyable!).

**Comment by geppo1988** [comment]
Thanks that awesome....we can build our pedalboard modifying a 3 euro
keyboard now :)

Il lun 2 ago 2021, 13:41 Christopher Cook ***@***.***> ha
scritto:
